Add support for Postgres 18

lfittl opened the pull request:
(against `17-latest` for the review, but for release we will instead rename the branch to `18-latest`)

## TODO

* [x] Fix "#include <xlocale.h>" error with CI (and presumably any Linux)
* [x] Fix deparser differences/crashes (`test/deparse`)
* [x] Fix PL/pgSQL parsing crash (`test/parse_plpgsql`)
* PL/pgSQL parsing
  * [x] Ensure we support all regular base type for PL/pgSQL parsing
  * ~~Consider adding an API for allowing to add custom types (fixes [row type handling](https://github.com/pganalyze/libpg_query/issues/159) and improves [composite type handling](https://github.com/pganalyze/libpg_query/issues/260))~~ (deferred)
  * ~~Consider adding a Protobuf output option for PL/pgSQL parse trees~~ (deferred)
* [x] Review added files to see if additional mocks can reduce them (many of them added because of type handling changes for PL/pgSQL)
* [x] Rebase after https://github.com/pganalyze/libpg_query/pull/325 is merged to fix [#254](https://github.com/pganalyze/libpg_query/issues/254)
* [x] Review if [#285](https://github.com/pganalyze/libpg_query/issues/285) is still applicable
* [x] Review fingerprinting (Postgres 18 changed some jumble logic)
* Review feedback
  * ~~Review if we can drop the LimitOption patch (05_limit_option_enum_value_default) -- its not clear what this is really adding at this point (AFAIK this was helpful for JSON output, or maybe the deparser?)~~ Changing this causes a lot of fingerprints to flip, which given the otherwise relatively low impact parse tree changes doesn't seem advisible
  * [x] Make add_parameter_name/add_dummy_return/build_row_from_vars exported (like other two static functions) -- or: figure out if instead of having compile_create_function_stmt which is mainly a copy of plpgsql_compile_callback we instead patch/mock plpgsql_compile_callback to do what we need (that would make it easier to rebase for new versions)
  * [x] Add a comment to GetSysCacheOid that explains why we always return `RECORDOID` for public namespace (and possibly note ideas on how we could extend it)
  * [x] Rebase on top of latest 18.X patch release (that may also allow taking out the XLOCALE_H changes)
  * [x] Consider patching out `object_aclcheck` from `interpret_function_parameter_list` instead of mocking the whole function
  * [x] Add a comment why LookupExplicitNamespace is mocked
  * [x] Review why we removed the return from `trgfn` (and if that function definition is from somewhere else, add a comment explaining it)

## TODO (non-release blocking)

* Review if [pg_query/#339](https://github.com/pganalyze/pg_query/issues/339) can be solved (that may be achievable through Postgres' mechanism for checking stack depth?)
* Review `strlcpy` issue encounted in [pglast testing](https://github.com/pganalyze/libpg_query/issues/304#issuecomment-3696031130)
* Review why we had `pg_catalog.XXX` types in PL/pgSQL before, and now its just `XXX` (see test diff)

lfittl committed:
Add support for Postgres 18

lfittl committed:
WIP: Add support for Postgres 18

lfittl committed:
Skip VariableSetStmt.jumble_args when fingerprinting

We may need to do this differently, but for now this shouldn't be
significant since it mainly relates to how the SET statement was worded,
not its intent.

lfittl committed:
Skip VariableSetStmt.jumble_args when fingerprinting

We may need to do this differently, but for now this shouldn't be
significant since it mainly relates to how the SET statement was worded,
not its intent.

lfittl committed:
Deparser: Update to accomodate 18 changes, add additional regression tests

lfittl committed:
Deparser: Update to accomodate 18 changes, add additional regression tests

lfittl committed:
Deparser: Pass opts by reference/pointer instead of copying it

This is a breaking change for callers, but seems more appropriate since
it avoids implicit shallow copying of the struct (which itself includes
pointer values, that would be retained in such a copy).

lfittl committed:
Deparser: Pass opts by reference/pointer instead of copying it

This is a breaking change for callers, but seems more appropriate since
it avoids implicit shallow copying of the struct (which itself includes
pointer values, that would be retained in such a copy).

lfittl committed:
Deparser: Rework comment handling for PG 18 multi-statement strings

Previously a subsequent statement in a multi-statement string would start
at the end of the prior statement, including any comments between. In
Postgres 18 this has changed, and such comments (and whitespace) are
excluded from the statements, the next statement starts at the first
non-whitespace (and non-comment) character.

Because of this we change the comments logic in the deparser: Previously
comments that had a location outside of the statement would be ignored
by deparseRawStmtOpts, which processes a single RawStmt. Instead, ensure
that all comments passed by the caller are emitted. This is a backwards
incompatible change for any direct callers of the deparser code.

For callers that call libpg_query instead of the deparser directly, e.g.
by using pg_query_deparse_protobuf_opts, nothing changes, as comments
are now split up before passing them down to the deparser code.

lfittl committed:
Deparser: Rework comment handling for PG 18 multi-statement strings

Previously a subsequent statement in a multi-statement string would start
at the end of the prior statement, including any comments between. In
Postgres 18 this has changed, and such comments (and whitespace) are
excluded from the statements, the next statement starts at the first
non-whitespace (and non-comment) character.

Because of this we change the comments logic in the deparser: Previously
comments that had a location outside of the statement would be ignored
by deparseRawStmtOpts, which processes a single RawStmt. Instead, ensure
that all comments passed by the caller are emitted. This is a backwards
incompatible change for any direct callers of the deparser code.

For callers that call libpg_query instead of the deparser directly, e.g.
by using pg_query_deparse_protobuf_opts, nothing changes, as comments
are now split up before passing them down to the deparser code.

lfittl committed:
Update LOCALE_T_IN_XLOCALE/WCSTOMBS_L_IN_XLOCALE to HAVE_XLOCALE_H

This was unified during the Postgres 18 cycle.

lfittl committed:
Update LOCALE_T_IN_XLOCALE/WCSTOMBS_L_IN_XLOCALE to HAVE_XLOCALE_H

This was unified during the Postgres 18 cycle.

lfittl committed:
Ensure HAVE_COPYFILE/HAVE_COPYFILE_H is not set outside of macOS

Previously this was only necessary to be handled on Windows, but Postgres
18 introduced optional use of copyfile.h in a few more places.

lfittl committed:
Ensure HAVE_COPYFILE/HAVE_COPYFILE_H is not set outside of macOS

Previously this was only necessary to be handled on Windows, but Postgres
18 introduced optional use of copyfile.h in a few more places.

lfittl committed:
Rework PL/pgSQL parsing to use proper type definitions

This is a substantial change to the hack-ish approach we've taken in
prior versions, instead of hacking our way to get the PL/pgSQL parser
to accept our constructs, imitate actual CREATE FUNCTION logic and
PL/pgSQL compilation for a function.

As part of a new release this now requires calling the
scripts/extract_pg_types.rb function to refresh built-in Postgres types.

In passing this improves the mock handling in the extract_source script.

lfittl committed:
Rework PL/pgSQL parsing to use proper type definitions

This is a substantial change to the hack-ish approach we've taken in
prior versions, instead of hacking our way to get the PL/pgSQL parser
to accept our constructs, imitate actual CREATE FUNCTION logic and
PL/pgSQL compilation for a function.

This patch currently adds a lot of files to libpg_query, and we may
want to take another look and see if some of them couldn't be mocked
higher up.

Additionally, this currently hardcodes a subset of all default Postgres
types, and only has partial array type support.

In passing this improves the mock handling in the extract_source script.

pyramation commented:
Excited for PG18!

Quick note: I’ve been working with PL/pgSQL parsing/deparsing in PG17 and ended up building a `hydratePlpgsqlAst()` helper. It walks the PL/pgSQL AST and parses embedded SQL expression strings (`PLpgSQL_expr.query`) into full SQL AST nodes, so downstream tooling can operate on structured ASTs instead of opaque strings.

I noticed a couple of unchecked boxes in PG18 and wanted to ask about the plan there—specifically whether `PLpgSQL_expr.query` will remain a string. That’s totally fine for me either way; I’m mainly asking so I can plan my work accordingly.

MBkkt commented:
@lfittl Hi, maybe it make sense to create branch for 18 without pl/pgsql support? Because it seems like this PR is stuck

lfittl committed:
Update to Postgres 18.4, patch interpret_function_parameter_list

lfittl committed:
PL/pgSQL sample: Add back return, but with required expression

We previously didn't use the parsing code correctly, hence this wasn't
flagged before and (AFAIR) an incorrect example.

lfittl committed:
Rework AllocSetDeleteFreeList to always delete both free lists

This was taking a context, but the free list is managed once per type
of context, not once per context. Instead, always free both free lists
currently in use in the function.

Fixes #285

lfittl committed:
Rework how PL/pgSQL compilation is done to rely on original code more

This reduces unique code we maintain in pg_query_parse_plpgsql.c, at the
downside of introducing additional mocks.

lfittl committed:
Fingerprinting: Rework alias/schema name handling to match Postgres 18

msepga reviewed (approved):
LGTM!

lfittl committed:
18.0.0 Release

In passing, omit singular authorship in README. Whilst I (Lukas) am
actively maintaining the library, there are significant co-contributors
and it seems easier to not list individual authors.

lfittl commented:
> Excited for PG18!
> 
> Quick note: I’ve been working with PL/pgSQL parsing/deparsing in PG17 and ended up building a `hydratePlpgsqlAst()` helper. It walks the PL/pgSQL AST and parses embedded SQL expression strings (`PLpgSQL_expr.query`) into full SQL AST nodes, so downstream tooling can operate on structured ASTs instead of opaque strings.
> 
> I noticed a couple of unchecked boxes in PG18 and wanted to ask about the plan there—specifically whether `PLpgSQL_expr.query` will remain a string. That’s totally fine for me either way; I’m mainly asking so I can plan my work accordingly.

I think it makes sense to have libpg_query optionally do the parsing of this as well (since we already have the parsemode, it'd be straightforward to do), but I ran out of time/focus on getting this in, so for now the PL/pgSQL output is the same as on the 17 branch.

I'm aiming to rework that a bit in the next weeks/months, but aiming to keep a compatibility mode for the current output.
